pkg/api/factory.go

```diff
@@ -93,36 +111,77 @@ func (f *apiFactory) getConfigMapAndSecret() (*v1.ConfigMap, *v1.Secret, error)
 		}
 	}
 
+	if errors.IsNotFound(err) {
+		return cm, secret, nil
+	}
 	return cm, secret, err
 }
 
-func (f *apiFactory) invalidateCache() {
-	f.lock.Lock()
-	defer f.lock.Unlock()
-	f.api = nil
+func (f *apiFactory) getConfigMapAndSecret(namespace string) (*v1.ConfigMap, *v1.Secret, error) {
+	cmLister := f.cmLister.ConfigMaps(namespace)
+	secretLister := f.secretLister.Secrets(namespace)
+
+	return f.getConfigMapAndSecretWithListers(cmLister, secretLister)
 }
 
 func (f *apiFactory) GetAPI() (API, error) {
+	apis, err := f.GetAPIsFromNamespace(f.Settings.DefaultNamespace)
+	if err != nil {
+		return nil, err
+	}
+	return apis[f.Settings.DefaultNamespace], nil
+}
+
+// For self-service notification, we need a map of apis which include api in the namespace and api in the setting's namespace
+func (f *apiFactory) GetAPIsFromNamespace(namespace string) (map[string]API, error) {
 	f.lock.Lock()
 	defer f.lock.Unlock()
-	if f.api == nil {
-		cm, secret, err := f.getConfigMapAndSecret()
-		if err != nil {
-			return nil, err
-		}
-		cfg, err := ParseConfig(cm, secret)
-		if err != nil {
-			return nil, err
-		}
-		getVars, err := f.InitGetVars(cfg, cm, secret)
-		if err != nil {
-			return nil, err
-		}
-		api, err := NewAPI(*cfg, getVars)
-		if err != nil {
-			return nil, err
+
+	apis := make(map[string]API)
+
+	// namespaces to look for notification configurations
+	namespaces := []string{namespace}
+	if !slices.Contains(namespaces, f.Settings.DefaultNamespace) {
+		namespaces = append(namespaces, f.Settings.DefaultNamespace)
+	}
+
+	for _, namespace := range namespaces {
+		if f.apiMap[namespace] == nil {
+			api, err := f.getApiFromNamespace(namespace)
+			if err != nil {
+				return nil, err
```

Comment:
Should we return error when we do not have API from any namespace? This is for scenario that only default namespace has notification configuration.

Reply:
Good call out I missed that